packages/nextjs/src/client/clientNormalizationIntegration.ts

```diff
@@ -2,30 +2,84 @@ import { rewriteFramesIntegration } from '@sentry/browser';
 import { defineIntegration } from '@sentry/core';
 
 export const nextjsClientStackFrameNormalizationIntegration = defineIntegration(
-  ({ assetPrefixPath }: { assetPrefixPath: string }) => {
+  ({
+    assetPrefix,
+    basePath,
+    rewriteFramesAssetPrefixPath,
+    experimentalThirdPartyOriginStackFrames,
+  }: {
+    assetPrefix?: string;
+    basePath?: string;
+    rewriteFramesAssetPrefixPath: string;
+    experimentalThirdPartyOriginStackFrames: boolean;
+  }) => {
     const rewriteFramesInstance = rewriteFramesIntegration({
       // Turn `<origin>/<path>/_next/static/...` into `app:///_next/static/...`
       iteratee: frame => {
-        try {
-          const { origin } = new URL(frame.filename as string);
-          frame.filename = frame.filename?.replace(origin, 'app://').replace(assetPrefixPath, '');
-        } catch (err) {
-          // Filename wasn't a properly formed URL, so there's nothing we can do
+        if (experimentalThirdPartyOriginStackFrames) {
+          // Not sure why but access to global WINDOW from @sentry/Browser causes hideous ci errors
+          // eslint-disable-next-line no-restricted-globals
+          const windowOrigin = typeof window !== 'undefined' && window.location ? window.location.origin : '';
+          // A filename starting with the local origin and not ending with JS is most likely JS in HTML which we do not want to rewrite
+          if (frame.filename?.startsWith(windowOrigin) && !frame.filename.endsWith('.js')) {
+            return frame;
+          }
+
+          if (assetPrefix) {
+            // If the user defined an asset prefix, we need to strip it so that we can match it with uploaded sourcemaps.
+            // assetPrefix always takes priority over basePath.
+            if (frame.filename?.startsWith(assetPrefix)) {
+              frame.filename = frame.filename.replace(assetPrefix, 'app://');
+            }
+          } else if (basePath) {
+            // If the user defined a base path, we need to strip it to match with uploaded sourcemaps.
+            // We should only do this for same-origin filenames though, so that third party assets are not rewritten.
+            try {
+              const { origin: frameOrigin } = new URL(frame.filename as string);
+              if (frameOrigin === windowOrigin) {
+                frame.filename = frame.filename?.replace(frameOrigin, 'app://').replace(basePath, '');
+              }
+            } catch (err) {
+              // Filename wasn't a properly formed URL, so there's nothing we can do
+            }
+          }
+        } else {
+          try {
+            const { origin } = new URL(frame.filename as string);
+            frame.filename = frame.filename?.replace(origin, 'app://').replace(rewriteFramesAssetPrefixPath, '');
```

Comment:
L: That's a minor code-style comment - I would move this logic (which is shorter) upward to improve readability.